adb_graphics/figure_builders.py

```diff
@@ -0,0 +1,280 @@
+
+    # pylint: disable=too-many-arguments,too-many-locals
+
+    '''
+    Function that creates plan-view maps, either a single panel, or
+    multipanel for a forecast ensemble. Can be used in parallel.
+
+    Input:
+
+      fhr        forecast hour
+      ds         xarray dataset from the grib file
+      level      the vertical level of the variable to be plotted
+                 corresponding to a key in the specs file
+      model      model name: rap, hrrr, hrrre, rrfs, rtma
+      spec       the dictionary of specifications for the given variable
+                 and level
+      variable   the name of the variable section in the specs file
+      workdir    output directory
+
+    Optional:
+      tile       the label of the tile being plotted
+    '''
+
+    fig, axes = set_figure(cla.model_name, cla.graphic_type, tile)
+
+    # set last_panel to send into DataMap for colorbar control
+    last_panel = False
+
+    # Declare the type of object depending on graphic type
+    map_class = maps.MultiPanelDataMap if cla.graphic_type == \
+        'enspanel' else maps.DataMap
+
+    for index, current_ax in enumerate(axes):
+
+        if current_ax is axes[-1]:
+            last_panel = True
+        mem = None
+        if cla.graphic_type == 'enspanel':
+            # Don't put data in the top left or bottom left panels.
+            if index in (0, 8):
+                current_ax.axis('off')
+
+            # Shenanigans to match ensemble member to panel index
+            mem = 0 if index == 4 else index
+            mem = mem if mem < 4 else index - 1
+            mem = mem if mem < 8 else index - 2
+
+        # Object to be plotted on the map in filled contours.
+        field = gribdata.fieldData(
+            ds=ds,
+            fhr=fhr,
+            filetype=cla.file_type,
+            level=level,
+            member=mem,
+            model=model,
+            short_name=variable,
+            )
+
+        try:
+            field.field
+        except errors.GribReadError:
+            print(f'Cannot find grib2 variable for {variable} at {level}. Skipping.')
+            return
+
+
+        map_fields = maps.MapFields(
+            fields_spec=spec,
+            main_field=field,
+            map_type=cla.graphic_type,
+            model=model,
+            tile=tile,
+            )
+
+        # Generate a map object
+        m = maps.Map(
+            airport_fn=AIRPORTS,
+            ax=current_ax,
+            grid_info=field.grid_info(),
+            model=model,
+            plot_airports=spec.get('plot_airports', True),
+            tile=tile,
+            )
+
+        # Send all objects (map_field, contours, hatches) to a DataMap object
+        dm = map_class(
+            map_fields=map_fields,
+            map_=m,
+            member=mem,
+            model_name=cla.model_name,
+            last_panel=last_panel
+            )
+
+        # Draw the map
+        if cla.graphic_type == 'enspanel':
```

Comment:
It was a design choice to add the title to the blank top left panel (panel 0) and observations for composite reflectivity to the bottom left (panel 8).